File: C#/714. Best Time to Buy and Sell Stock with Transaction Fee.cs
public class Solution {
    public int MaxProfit(int[] prices, int fee) {
        int n=prices.Length;
        int[][] dp=new int[n][];
        for(int i=0; i<dp.Length; i++){
            dp[i] = new int[2];
            dp[i][0] = 0;
            dp[i][1] = 0;
        }
        return helper(0,1,fee,prices,dp);
    }
    public int helper(int i,int buy,int fee,int[] prices,int[][] dp){
        if(i>=prices.Length){
            return 0;
        }
        if(dp[i][buy]!=0){
            return dp[i][buy];
        }
        if(buy!=0){
            dp[i][buy]=Math.Max(-prices[i]+helper(i+1,0,fee,prices,dp),helper(i+1,1,fee,prices,dp));
        }
        else{
            dp[i][buy]=Math.Max(prices[i]-fee+helper(i+1,1,fee,prices,dp),helper(i+1,0,fee,prices,dp));
        }
        return dp[i][buy];
    }
}

File: C++/714. Best Time to Buy and Sell Stock with Transaction Fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n=prices.size();
        vector<vector<int>> dp(n,vector<int>(2,-1));
        return helper(0,1,fee,prices,dp);
    }
    int helper(int i,int buy,int fee,vector<int>& prices,vector<vector<int>>& dp){
        if(i>=prices.size()){
            return 0;
        }
        if(dp[i][buy]!=-1){
            return dp[i][buy];
        }
        if(buy!=0){
            dp[i][buy]=max(-prices[i]+helper(i+1,0,fee,prices,dp),helper(i+1,1,fee,prices,dp));
        }
        else{
            dp[i][buy]=max(prices[i]-fee+helper(i+1,1,fee,prices,dp),helper(i+1,0,fee,prices,dp));
        }
        return dp[i][buy];
    }
};
